alibi/prototypes/protoselect.py

```diff
@@ -0,0 +1,543 @@
+        # [NY, L]. For every row `y` and every column `l`, we compute how many new instances belonging to all the
+        # other classes different then `l` will be covered if we add the prototype `y`.
+        delta_nu_summed = np.sum(delta_nu_all, axis=-1)
+        # compute the tradeoff score - each prototype tries to cover as many new elements as possible
+        # belonging to the same class, while trying to avoid covering elements belonging to another class
+        scores_all = delta_xi_summed - delta_nu_summed - self.lbd
+
+        # add progressing bar if `verbose=True`
+        generator = range(num_prototypes)
+        if self.verbose:
+            generator = tqdm(generator)
+
+        for _ in generator:
+            j = np.array(list(available_indices)).astype(np.int32)
+            scores = scores_all[j]
+
+            # stopping criterion. The number of the returned prototypes might be lower than
+            # the number of requested prototypes
+            if np.all(scores < 0):
+                break
+
+            # find the index `i` of the best prototype and the class `l` that it covers
+            row, col = np.unravel_index(np.argmax(scores), scores.shape)
+            i, l = j[row.item()], col.item()  # noqa: E741
+
+            # update the score
+            covered = np.sum(delta_xi_all[:, l, B[i].astype(bool)], axis=-1)
+            delta_xi_all[:, l, B[i].astype(bool)] = 0
+            delta_xi_summed[:, l] -= covered
+            scores_all[:, l] -= covered
+
+            # add prototype to the corresponding list according to the class label `l` that it covers
+            # and remove the index `i` from list of available indices
+            protos[l].append(i)
+            available_indices.remove(i)
+
+        return self._build_explanation(protos)
+
+    def _build_explanation(self, protos: Dict[int, List[int]]) -> Explanation:
+        """
+        Helper method to build `Explanation` object.
+        """
+        data = deepcopy(DEFAULT_DATA_PROTOSELECT)
+        data['prototypes_indices'] = np.concatenate(list(protos.values())).astype(np.int32)
+        data['prototypes_labels'] = np.concatenate([[l] * len(protos[l])
+                                                    for l in protos]).astype(np.int32)  # noqa: E741
+        data['prototypes'] = self.Y[data['prototypes_indices']]
+        return Explanation(meta=self.meta, data=data)
+
+    def save(self, path: Union[str, os.PathLike]) -> None:
+        super().save(path)
+
+    @classmethod
+    def load(cls, path: Union[str, os.PathLike]) -> "Explainer":  # type: ignore[override]
+        return super().load(path, predictor=None)
+
+
+def _helper_protoselect_euclidean_1knn(explainer: ProtoSelect,
+                                       num_prototypes: int,
+                                       eps: float,
+                                       preprocess_fn: Optional[Callable[[np.ndarray], np.ndarray]] = None
```

Comment:
`preprocess_fn` is currently unused

Reply:
Forgot to remove it after removing some unnecessary re-computation.